src/storage-types/src/sources/mysql.rs

```diff
@@ -159,3 +173,259 @@ impl RustType<ProtoMySqlSourceDetails> for MySqlSourceDetails {
+/// TODO(roshan): Add compatibility for MySQL 8.3 'Tagged' GTIDs
+pub fn gtid_set_frontier(gtid_set_str: &str) -> Result<Antichain<GtidPartition>, io::Error> {
+    let mut partitions = Antichain::new();
+    for gtid_str in gtid_set_str.split(',') {
+        if gtid_str.is_empty() {
+            continue;
+        };
+        let (uuid, intervals) = gtid_str.split_once(':').ok_or_else(|| {
+            std::io::Error::new(
+                std::io::ErrorKind::InvalidData,
+                format!("invalid gtid: {}", gtid_str),
+            )
+        })?;
+
+        let uuid = Uuid::parse_str(uuid).map_err(|e| {
+            io::Error::new(
+                io::ErrorKind::InvalidData,
+                format!("invalid uuid in gtid: {}: {}", gtid_str, e),
+            )
+        })?;
+
+        // From the MySQL docs:
+        // "When GTID sets are returned from server variables, UUIDs are in alphabetical order,
+        // and numeric intervals are merged and in ascending order."
+        // For our purposes, we need to ensure that all intervals are consecutive which means there
+        // should be at most one interval per GTID.
+        // TODO: should this same restriction be done when parsing a @@GTID_PURGED value? In that
+        // case the intervals might not be guaranteed to be consecutive, depending on how purging
+        // is implemented?
+        let mut intervals = intervals.split(':');
+        let end = match (intervals.next(), intervals.next()) {
+            (Some(interval_str), None) => {
+                let mut vals_iter = interval_str.split('-').map(str::parse::<u64>);
+                let start = vals_iter
+                    .next()
+                    .ok_or_else(|| {
+                        io::Error::new(
+                            io::ErrorKind::InvalidData,
+                            format!("couldn't parse int: {}", interval_str),
+                        )
+                    })?
+                    .map_err(|e| {
+                        io::Error::new(
+                            io::ErrorKind::InvalidData,
+                            format!("couldn't parse int: {}: {}", interval_str, e),
+                        )
+                    })?;
+                match vals_iter.next() {
+                    Some(Ok(end)) => end,
+                    None => start,
+                    _ => {
+                        return Err(io::Error::new(
+                            io::ErrorKind::InvalidData,
+                            format!("invalid gtid interval: {}", interval_str),
+                        ))
+                    }
+                }
+            }
+            _ => {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("gtid with non-consecutive intervals found! {}", gtid_str),
+                ))
+            }
+        };
+        // Insert a partition representing the 'next' GTID that might be seen from this source
```

Comment:
Since the uuids are given to us in alphabetical order it is a bit easier to produce the gaps as we go:

```diff
diff --git a/src/storage-types/src/sources/mysql.rs b/src/storage-types/src/sources/mysql.rs
index e3888707fc..514f9a124f 100644
--- a/src/storage-types/src/sources/mysql.rs
+++ b/src/storage-types/src/sources/mysql.rs
@@ -321,6 +321,7 @@ impl SourceTimestamp for GtidPartition {
 /// TODO(roshan): Add compatibility for MySQL 8.3 'Tagged' GTIDs
 pub fn gtid_set_frontier(gtid_set_str: &str) -> Result<Antichain<GtidPartition>, io::Error> {
     let mut partitions = Antichain::new();
+    let gap_lower = Some(Uuid::nil());
     for gtid_str in gtid_set_str.split(',') {
         if gtid_str.is_empty() {
             continue;
@@ -383,49 +384,20 @@ pub fn gtid_set_frontier(gtid_set_str: &str) -> Result<Antichain<GtidPartition>,
                 ))
             }
         };
+        if let Some(gap_upper) = uuid.backward_checked(1) {
+            let gap_lower = gap_lower.expect("uuids are in alphabetical order");
+            partitions.insert(GtidPartition::new_range(gap_lower, gap_upper, GTIDState::Absent));
+        }
+        gap_lower = uuid.forward_checked(1);
         // Insert a partition representing the 'next' GTID that might be seen from this source
         partitions.insert(GtidPartition::new_singleton(
             uuid,
             GTIDState::Active(NonZeroU64::new(end + 1).unwrap()),
         ));
     }
-
-    // Fill in gaps in the GTID set with Absent values for all future source-ids we may see
-    let mut future_partitions = vec![];
-
-    let mut prev = Uuid::nil();
-    let mut push_last = true;
-    for partition in partitions.iter().sorted_by(|a, b| {
-        a.interval()
-            .singleton()
-            .unwrap()
-            .cmp(b.interval().singleton().unwrap())
-    }) {
-        let part = partition.interval().singleton().unwrap();
-        assert!(prev < *part, "duplicate GTID found in GTID set");
-
-        // Create a partition representing all the UUIDs in the gap between the previous one and this one
-        if let Some(end) = part.backward_checked(1) {
-            future_partitions.push(GtidPartition::new_range(prev, end, GTIDState::Absent));
-        }
-
-        // Set prev to the UUID + 1 so the next gap starts after this one
-        if let Some(next) = part.forward_checked(1) {
-            prev = next;
-        } else {
-            // part was UUID::MAX so we don't need to add a future partition
-            push_last = false;
-            break;
-        };
+    if let Some(gap_lower) = gap_lower {
+        partitions.insert(GtidPartition::new_range(lower, Uuid::max(), GTIDState::Absent));
     }
-    if push_last {
-        future_partitions.push(GtidPartition::new_range(
-            prev,
-            Uuid::maximum(),
-            GTIDState::Absent,
-        ));
-    };
 
-    partitions.extend(future_partitions.into_iter());
     Ok(partitions)
 }
 ```